awscrt/websocket.py

```diff
@@ -259,7 +277,18 @@ def _on_connection_setup(
                 cbdata.websocket.close()
 
     def _on_connection_shutdown(self, error_code):
-        pass  # TODO
+        cbdata = OnConnectionShutdownData()
+        if error_code:
+            cbdata.exception = awscrt.exceptions.from_code(error_code)
+
+        # Do not let exceptions from the user's callback bubble up any further.
+        try:
+            self._on_connection_shutdown_cb(cbdata)
+        except Exception:
+            print("Exception in WebSocket on_connection_shutdown callback", file=sys.stderr)
+            sys.excepthook(*sys.exc_info())
+
+        self._shutdown_event.set()
```

Comment:
Why not use future and just have the exceptions raised from the future if happens? So that, we don't need to have a callback for the exception and the event for wait.

Reply:
tldr: I think it's still a good idea to have a shutdown callback, that they can pass along with every other callback in the `websocket.connect()` function

Longer explanation:
I'm trying to avoid the classic creation-race-condition we keep running into when we mix callback-based APIs with stuff that's more synchronous in nature.

That's why `connect()` doesn't return a `Future[Websocket]`. The user can only get their hands on the websocket via the `on_connection_setup(...)` callback.

So if we only had a `shutdown_future`, but the user wanted to react to shutdown in an event-based way with a callback (instead of synchronously waiting on `shutdown_future.result()`) their code would need to look like this:
```py
class MyWebsocketHandler:
    def on_connection_setup(self, setup_data):
        self.websocket = setup_data.websocket
        # set shutdown callback here
        self.websocket.shutdown_future.add_done_callback(self.on_connection_shutdown)

    def on_connection_shutdown(self, shutdown_data):
        ...

    def on_incoming_frame_payload(self, incoming_data):
        ...


def main():
    ws_handler = MyWebsocketHandler()
    # but all my other callbacks are set here
    awscrt.websocket.connect(host, port,
        on_connection_setup=ws_handler.on_connection_setup,
        on_incoming_frame_payload=ws_handler.on_incoming_frame_payload)
```

I think the user would prefer to pass all their callbacks in one place, like:
```py
def main():
    ws_handler = MyWebsocketHandler()
    # but all my other callbacks are set here
    awscrt.websocket.connect(host, port,
        on_connection_setup=ws_handler.on_connection_setup,
        on_connection_shutdown=ws_handler.on_connection_shutdown,
        on_incoming_frame_payload=ws_handler.on_incoming_frame_payload)
```

Reply:
I flip-flopped on whether to return `Event`, `Future`, or `None` from `close()`....

we just chatted and you mentioned how offering multiple ways of learning about something can confuse the user. I agree. Let's keep it simple and not return anything from `close()`. We could always add something later